src/main/java/org/jenkinsci/remoting/util/VersionNumber.java

```diff
@@ -0,0 +1,579 @@
+            if (followedByDigit && value.length() == 1) {
+                // a1 = alpha-1, b1 = beta-1, m1 = milestone-1
+                switch (value.charAt(0)) {
+                    case 'a':
+                        value = "alpha";
+                        break;
+                    case 'b':
+                        value = "beta";
+                        break;
+                    case 'm':
+                        value = "milestone";
+                        break;
+                    default:
+                        /* fall through? */
+                }
+            }
+            this.value = ALIASES.getProperty(value, value);
+        }
+
+        public int getType() {
+            return STRING_ITEM;
+        }
+
+        public boolean isNull() {
+            return (comparableQualifier(value).compareTo(RELEASE_VERSION_INDEX) == 0);
+        }
+
+        /**
+         * Returns a comparable for a qualifier.
+         * <p/>
+         * This method both takes into account the ordering of known qualifiers as well as lexical ordering for unknown
+         * qualifiers.
+         * <p/>
+         * just returning an Integer with the index here is faster, but requires a lot of if/then/else to check for -1
+         * or QUALIFIERS.size and then resort to lexical ordering. Most comparisons are decided by the first character,
+         * so this is still fast. If more characters are needed then it requires a lexical sort anyway.
+         *
+         * @param qualifier
+         * @return
+         */
+        public static String comparableQualifier(String qualifier) {
+            int i = _QUALIFIERS.indexOf(qualifier);
+
+            return i == -1 ? _QUALIFIERS.size() + "-" + qualifier : String.valueOf(i);
+        }
+
+        public int compare(Item item) {
+            if (item == null) {
+                // 1-rc < 1, 1-ga > 1
+                return comparableQualifier(value).compareTo(RELEASE_VERSION_INDEX);
+            }
+            switch (item.getType()) {
+                case INTEGER_ITEM:
+                    return -1; // 1.any < 1.1 ?
+
+                case STRING_ITEM:
+                    if (item instanceof StringItem) {
+                        return comparableQualifier(value).compareTo(comparableQualifier(((StringItem) item).value));
+                    }
+
+                case LIST_ITEM:
+                    return -1; // 1.any < 1-1
+
+                case WILDCARD_ITEM:
+                    return -1;
+
+                default:
+                    throw new RuntimeException("invalid item: " + item.getClass());
+            }
+        }
+
+        public String toString() {
+            return value;
+        }
+    }
+
+    /**
+     * Represents a version list item. This class is used both for the global item list and for sub-lists (which start
+     * with '-(number)' in the version specification).
+     */
+    private static class ListItem extends ArrayList<Item> implements Item {
```

Comment:
I would not suggest extending collection classes. You can just delegate to an instance of one.